### PR TITLE
Add observer instrument

### LINKED ISSUE #531: Implement Observer instrument and corresponding aggregation.
We have an instrument named Observer already, but it does not have any 'callback' mechanisms implemented. This issue is to track implementing an Observer as per the specs.

https://github.com/open-telemetry/oteps/blob/master/text/0072-metric-observer.md

### PR DESCRIPTION
**Pull request by cijothomas**
Fix #531 
Observer instruments do not have access to context, and they also don't expose public bound instrument, as metrics are observed via callbacks only.
Observer uses LastValue aggregator.

Additional Note: Observer instrument part is implemented without generics. i.e Int64Observer and DoubleObserver are separate classes. 
If all okay, a follow up PR would replace generics for all other instruments as well.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/DoubleObserverMetricSdk.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    internal class DoubleObserverMetricSdk : DoubleObserverMetric
+    {
+        private readonly IDictionary<LabelSet, DoubleObserverMetricHandleSdk> observerHandles = new ConcurrentDictionary<LabelSet, DoubleObserverMetricHandleSdk>();
+        private string metricName;
+        private Action<DoubleObserverMetric> callback;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/DoubleObserverMetricSdk.cs]:
```suggestion
        private readonly Action<DoubleObserverMetric> callback;
```

File: src/OpenTelemetry/Metrics/DoubleObserverMetricSdk.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    internal class DoubleObserverMetricSdk : DoubleObserverMetric
+    {
+        private readonly IDictionary<LabelSet, DoubleObserverMetricHandleSdk> observerHandles = new ConcurrentDictionary<LabelSet, DoubleObserverMetricHandleSdk>();
+        private string metricName;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/DoubleObserverMetricSdk.cs]:
```suggestion
        private readonly string metricName;
```

File: src/OpenTelemetry/Metrics/Int64ObserverMetricSdk.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    internal class Int64ObserverMetricSdk : Int64ObserverMetric
+    {
+        private readonly IDictionary<LabelSet, Int64ObserverMetricHandleSdk> observerHandles = new ConcurrentDictionary<LabelSet, Int64ObserverMetricHandleSdk>();
+        private string metricName;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/Int64ObserverMetricSdk.cs]:
```suggestion
        private readonly string metricName;
```

File: src/OpenTelemetry/Metrics/Int64ObserverMetricSdk.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        private Action<Int64ObserverMetric> callback;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/Int64ObserverMetricSdk.cs]:
```suggestion
        private Action<Int64ObserverMetric> callback;
```

File: src/OpenTelemetry/Metrics/MeterSdk.cs [hunk 5]
```diff
@@ -115,11 +127,21 @@ public void Collect()
...
+                foreach (var doubleObserver in this.doubleObservers)
                 {
-                    var metricName = longObserver.Key;
-                    var measureInstrument = longObserver.Value;
-                    foreach (var handle in measureInstrument.GetAllHandles())
+                    var metricName = doubleObserver.Key;
+                    var observerInstrument = doubleObserver.Value;
+                    try
+                    {
+                        // TODO: Decide if we want to enforce a timeout.
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/MeterSdk.cs]: Can you raise an issue to track this `TODO` please.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterSdk.cs]: done

**Comment by cijothomas** [comment]
@MikeGoldsmith Addressed comments. Please re-review.

**Comment by cijothomas** [comment]
@SergeyKanzhelev Please merge.

**Pull request merged**